Sources/XCTest/XCTestCaseType.swift

```diff
@@ -0,0 +1,119 @@
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//
+//  XCTestCaseType.swift
+//  Base protocol (and extension with default methods) for test cases
+//
+
+public typealias TestMethod = XCTestCaseType throws -> Void
+
+public protocol XCTestCaseType {
+    init()
+
+    func setUp()
+    func tearDown()
+
+    // In the Objective-C version of XCTest, it is possible to discover all tests when the test is executed by asking the runtime for all methods and looking for the string "test". In Swift, we ask test providers to tell us the list of tests by implementing this property.
+    static var allTests: [(String, TestMethod)] { get }
+}
+
+/// Wrapper function allowing a static reference to a test method to fit 
+/// the signature required by XCTestCaseType.allTests
+public func test<T: XCTestCaseType>(testFunc: T -> () throws -> Void) -> TestMethod {
+    return { testCaseType in
+        guard let testCase: T = testCaseType as? T else {
+            fatalError("Attempt to invoke test on class \(T.self) with incompatible instance type \(testCaseType.dynamicType)")
+        }
+
+        try testFunc(testCase)()
+    }
+}
```

Comment:
Kudos on the Swift-fu here! :key: 

One suggestion for this pull request: by merging this we broaden the API surface area considerably. In addition to the current API, users will have to understand `XCTestCaseType` and the `test()` function. I think it's very important that the documentation for each of these types is approachable and user-friendly.

For example, the documentation for this function is very clear and explains the function's purpose. However, a user completely new to swift-corelibs-xctest may not understand where and why to use it. I'd like to consider providing an example, like so:

```swift
/// Wraps a static reference to an instance method on a test case such that it can be run by XCTest.
/// Use this when specifying your test methods in `XCTestCaseType.allTests()`:
///
///     class MyTestCase: XCTestCase {
///         static var allTests: [(String, XCTestCaseType throws -> Void)] {
///             ("testSomething", test(testSomething)),
///         }
///         func testSomething() {
///             // ...
///         }
///     }
```

Similarly, it might be nice to provide documentation for `XCTestCaseType`, to answer the question: "where should I use this public protocol?"